minorminer/utils/zephyr/qfloor.py

```diff
@@ -0,0 +1,402 @@
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+# ================================================================================================
+
+
+from __future__ import annotations
+
+from collections import defaultdict, namedtuple
+from itertools import product
+from functools import cached_property
+from typing import Callable, Iterable, Iterator
+
+from minorminer.utils.zephyr.coordinate_systems import *
+from minorminer.utils.zephyr.node_edge import ZEdge, ZNode, ZShape, EdgeKind
+from minorminer.utils.zephyr.plane_shift import PlaneShift
+
+Dim_2D = namedtuple("Dim_2D", ["Lx", "Ly"])
+UWJ = namedtuple("UWJ", ["u", "w", "j"])
+ZSE = namedtuple("ZSE", ["z_start", "z_end"])
+
+
+class QuoTile:
+    """Initializes a 'QuoTile' object with zns.
+
+    Args:
+        zns (Iterable[ZNode]): The ZNodes the tile contains
+
+    Example:
+    .. code-block:: python
+        >>> from minorminer.utils.zephyr.node_edge import ZNode
+        >>> from minorminer.utils.zephyr.qfloor import QuoTile
+        >>> ccoords = [(k, 1+k) for k in range(4)] + [(1+k, k) for k in range(4)]
+        >>> zns = [ZNode(coord=ccoord) for ccoord in ccoords]
+        >>> tile = QuoTile(zns)
+        >>> print(f"{tile.edges() = }\n{tile.seed = }\n{tile.shifts = }")
+    """
+
+    def __init__(
+        self,
+        zns: Iterable[ZNode],
+    ) -> None:
+        if len(zns) == 0:
+            raise ValueError(f"Expected zns to be non-empty, got {zns}")
+        for zn in zns:
+            if not zn.is_quo():
+                raise ValueError(f"Expected elements of {zns} to be quotient, got {zn}")
+        zns_shape = {zn.shape for zn in zns}
+        if len(zns_shape) != 1:
+            raise ValueError(
+                f"Expected all elements of zns to have the same shape, got {zns_shape}"
+            )
+        temp_zns = sorted(list(set(zns)))
+        seed_convert = temp_zns[0].convert_to_z
+        zns_result = []
+        for zn in temp_zns:
+            zn.convert_to_z = seed_convert
+            zns_result.append(zn)
+        self._zns: list[ZNode] = zns_result
+
+    @property
+    def zns(self) -> list[ZNode]:
+        """Returns the sorted list of ZNodes the tile contains"""
+        return self._zns
+
+    @cached_property
+    def index_zns(self) -> dict[int, ZNode]:
+        return {i: zn for i, zn in enumerate(self._zns)}
+
+    @property
+    def seed(self) -> ZNode:
+        """Returns the smallest ZNode in Zns"""
```

Comment:
(minor) Missing punctuation after line in a few of these methods.
```suggestion
        """Returns the smallest ZNode in Zns."""
```